tools/hidl/AccessMapGenerator.hpp

```diff
@@ -0,0 +1,1003 @@
+        }
+        postEntry();
+
+        return true;
+    }
+
+protected:
+    const ScopeMapper& m_scopeMapper;
+    const LevelMapper& m_levelMapper;
+    const Reporter& m_reporter;
+    Variables m_variables;
+    Scopes m_scopes;
+    bool m_requireSecurityAffirmation = true;
+    bool m_useDefaultSecurity = false;
+    Terms m_defaultSecurityTerms;
+
+private:
+    void normalizeAndSetValue(std::string& str, const char* value, size_t start = 0, size_t end = std::string::npos) const
+    {
+        str.clear();
+        for (size_t idx = start; idx < end && value[idx] != '\0'; idx++)
+        {
+            const char& ch = value[idx];
+            if (!isspace(ch) && ch != '"')
+                str.push_back(ch);
+        }
+    }
+
+    void tokenize(int srcScope, const std::string& fragment, std::list<std::string>& tokens)
+    {
+        debug("tokenizing %s fragment \"%s\"", m_scopeMapper(srcScope), fragment.c_str());
+        static const std::string ignoredCharacters(" \t\n\r\"");
+        std::string token;
+        bool done = false;
+        bool inVariable = false;
+        std::string variable;
+        size_t idx = 0;
+
+        while (!done)
+        {
+            const char& ch = fragment[idx++];
+            switch (ch)
+            {
+            // token delimiters
+            case '\0':
+                done = true;
+                // fall through
+            case ',':
+                if (inVariable)
+                {
+                    warning("%s fragment '%s' contains invalid variable markup (incomplete variable reference '%s')", m_scopeMapper(srcScope), fragment.c_str(), variable.c_str());
+                    token.append("${").append(variable);
+                    variable.clear();
+                    inVariable = false;
+                }
+                if (!token.empty())
+                {
+                    tokens.push_back(token);
+                    token.clear();
+                }
+                break;
+
+            // variable replacement
+            case '$':
+                while (isspace(fragment[idx])) idx++;
+                if ('{' == fragment[idx])
+                {
+                    idx++;
+                    if (!inVariable)
+                    {
+                        inVariable = true;
+                    }
+                    else
+                    {
+                        error("%s fragment '%s' contains invalid variable markup (nested variable reference)", m_scopeMapper(srcScope), fragment.c_str());
+                        token.clear();
+                        variable.clear();
+                        inVariable = false;
+                        while (fragment[idx] != '\0' && fragment[idx] != ',') idx++;
+                        done = '\0' == fragment[idx];
+                        continue;
+                    }
+                }
+                else
+                {
+                    token.push_back(ch);
```

Comment:
It is valid to have a fragment like '$something'- what does that mean? Aren't the {} required part of the variable substitution syntax?

Reply:
It means nothing to the generator, but it may mean something to the author that used it. It wasn't prohibited before and is not prohibited now.